crates/red_knot/src/lint.rs

```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
     Ok(())
 }
 
+fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
+    // TODO possibly we should have a special marker on the real typing module (from typeshed) so
+    // if you have your own "typing" module in your project, we don't consider it THE typing module
+    let Some(typing_module) = context.db.resolve_module(ModuleName::new("typing"))? else {
+        debug_assert!(false, "typing module should always exist");
+        return Ok(());
+    };
+    let typing_file = context.db.module_to_file(typing_module)?;
+    let typing_table = context.db.symbol_table(typing_file)?;
+    let Some(typing_override) = typing_table.root_symbol_id_by_name("override") else {
+        debug_assert!(false, "typing.override should exist");
+        return Ok(());
+    };
+
+    // TODO we should maybe index definitions by type instead of iterating all, or else iterate all
+    // just once, match, and branch to all lint rules that care about a type of definition
+    for (symbol, definition) in context.symbols().all_definitions() {
+        if !matches!(definition, Definition::FunctionDef(_)) {
+            continue;
+        }
+        let ty = context
+            .db
+            .infer_definition_type(context.file_id, symbol, definition.clone())?;
```

Comment:
We can't retrieve definition from symbol because the relationship is one(symbol)-to-many(definitions).

I would like to solve this issue by making `Definition` `Copy`, if I can resolve whatever problem stopped me from doing that already.

I think we could also pass `Definition` by reference here. Whether `Copy` or reference makes more sense depends on whether we expect `Definition` to get any bigger, which I don't.

Reply:
I think interning Definitions is probably a good direction here; I added a TODO for that.